presto-main/src/main/resources/webapp/src/components/SQLInput.jsx

```diff
@@ -0,0 +1,237 @@
+        const parser = new SqlBaseParser(new antlr4.CommonTokenStream(new SqlBaseLexer(new UpperCaseCharStream(sqlInfo.sql))));
+        const selectDetector = new SelectListener();
+        parser.addErrorListener(syntaxError);
+        antlr4.tree.ParseTreeWalker.DEFAULT.walk(selectDetector, parser.statement());
+        //check syntax error
+        if (syntaxError.error) {
+            errorHandler(syntaxError.error);
+            return false;
+        }
+        if (selectDetector.isSelect) {
+            if (typeof (selectDetector.limit) === 'string' || selectDetector.limit > 100) {
+                sqlInfo.sql = sqlInfo.sql.replace(limitRE, 'limit 100');
+            } else if (selectDetector.fetchFirstNRows > 100) {
+                sqlInfo.sql = sqlInfo.sql.replace(fetchFirstRE, 'fetch first 100 rows only');
+            } else if (selectDetector.limit === -1 && selectDetector.fetchFirstNRows === -1) {
+                sqlInfo.sql += ' limit 100';
+            }
+        }
+    } catch (err) {
+        if (errorHandler) {
+            errorHandler(err);
+        }
+        return false;
+    }
+    return true;
+}
+
+// SQL Input, including sql text, catalog(optional), and schema(optional)
+export function SQLInput({ handleSQL, show, enabled, initialSQL, errorHandler }) {
+    const sqlInfo = React.useRef({ sql: undefined, catalog: undefined, schema: undefined });
+    const [data, setData] = React.useState({ catalogs: [], schemas: [] });
+    const [code, setCode] = React.useState(initialSQL);
+
+    const checkValue = () => {
+        if (code.length > 0) {
+            sqlInfo.current.sql = code;
+            if (sqlCleaning(sqlInfo.current, errorHandler)) {
+                setCode(sqlInfo.current.sql);
+                handleSQL(sqlInfo.current);
+            }
+        }
+    };
+
+    async function setCatalog(catalog) {
+        sqlInfo.current.catalog = catalog;
+        // also reset schema
+        setData({ ...data, schemas: [] });
+        // fetch schemas from the catalog
+        const client = createClient();
+        try {
+            const schemas = await client.getSchemas(catalog);
+            setData({ ...data, schemas });
+        } catch (e) {
+            // use this to report the issue
+            console.error(e);
+        }
+    }
+
+    function setSchema(schema) {
+        sqlInfo.current.schema = schema;
+        setData({ ...data });
+    }
+
+    React.useEffect(() => {
+        //fetch catalogs:
+        async function getCatalogs() {
+            const client = createClient();
+            try {
+                const catalogs = await client.getCatalogs();
+                setData({ ...data, catalogs });
+            } catch (e) {
+                // use this to report the issue
+                console.error(e);
+            }
+        }
+        getCatalogs();
+    }, [initialSQL]);
+
+    return (
+        <div className={clsx(!show && 'hide')}>
```

Comment:
same as above

Reply:
I just ran a simple benchmark to compare the `Strings (variadic)` vs `Objects`. `Strings (variadic)` is faster.
I guess the `Objects` way has better readability. But for one class case, I vote for performance and simplicity.

Reply:
got, makes a lot of sense @yhwang 👍  yeah one class might be a little overkill, I was just inclining on readability for most of my comments. Thanks for taking a look!